diff --git a/3rdparty/assimp/code/Importer.cpp b/3rdparty/assimp/code/Importer.cpp
index 97f97c9..dec0d61 100644
--- a/3rdparty/assimp/code/Importer.cpp
+++ b/3rdparty/assimp/code/Importer.cpp
@@ -908,6 +908,8 @@ void WriteLogOpening(const std::string& file)
         << " ppc32"
 #elif defined(ASSIMP_BUILD_ARM_32BIT_ARCHITECTURE)
         << " arm32"
+#elif defined(ASSIMP_BUILD_MIPS_32BIT_ARCHITECTURE)
+	<< " mips32"
 #else
 #    error unknown architecture
 #endif
diff --git a/3rdparty/assimp/include/aiDefines.h b/3rdparty/assimp/include/aiDefines.h
index b6c5ff9..6b29289 100644
--- a/3rdparty/assimp/include/aiDefines.h
+++ b/3rdparty/assimp/include/aiDefines.h
@@ -242,6 +242,8 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #  define ASSIMP_BUILD_PPC_32BIT_ARCHITECTURE
 #       elif defined(__arm__)
 #               define ASSIMP_BUILD_ARM_32BIT_ARCHITECTURE
+#	elif defined(__mips__)
+#		define ASSIMP_BUILD_MIPS_32BIT_ARCHITECTURE
 # else
 #  error unknown architecture
 # endif
